python/cuml/_thirdparty/sklearn/preprocessing/_imputation.py

```diff
@@ -379,27 +379,40 @@ def _dense_fit(self, X, strategy, missing_values, fill_value):
             count_missing_values = mask.sum(axis=0)
             n_elems = X.shape[0] - count_missing_values
             mean = np.nansum(X, axis=0)
-            mean -= (count_missing_values * missing_values)
+            if not np.isnan(missing_values):
+                mean -= (count_missing_values * missing_values)
             mean /= n_elems
             return mean
 
         # Median
         elif strategy == "median":
+            if X.size == 0:
+                return np.full(X.shape[1], np.nan)
+            X_sorted = X.copy()
+            if not np.isnan(missing_values):
+                # If nan is not the missing value, any column with nans should
+                # have a median of nan
+                nan_cols = np.any(np.isnan(X), axis=0)
+                X_sorted[mask] = np.nan
+            else:
+                nan_cols = np.full(X.shape[1], False)
+            # nans are always sorted to end of array
+            X_sorted = np.sort(X_sorted, axis=0)
+
             count_missing_values = mask.sum(axis=0)
+            # Ignore missing values in determining "halfway" index of sorted
+            # array
             n_elems = X.shape[0] - count_missing_values
-            middle, is_odd = np.divmod(n_elems, 2)
-            is_odd = is_odd.astype(np.bool)
-            middle += count_missing_values
-            X_sorted = X.copy()
-            X_sorted[mask] = np.nan
-            X_sorted = np.sort(X, axis=0)
-            median = np.empty(X.shape[1], dtype=X.dtype)
-            wis_odd = np.argwhere(is_odd).squeeze()
-            wnot_odd = np.argwhere(~is_odd).squeeze()
-            median[wis_odd] = X_sorted[middle[wis_odd], wis_odd]
-            elm1 = X_sorted[middle[wnot_odd]-1, wnot_odd]
-            elm2 = X_sorted[middle[wnot_odd], wnot_odd]
-            median[wnot_odd] = (elm1 + elm2) / 2.
+
+            # If no elements remain after removing missing value, median for
```

Comment:
could this median logic be broken out to a helper function?

Reply:
What is the appropriate file for this code? Maybe `cuml/thirdparty_adapters/adatpers.py` (don't have any mathematical function for now though) or new file `cuml/prims/cupy_helper.py`?

Reply:
Yep, I think it makes good sense to break this out. I'll put it in `cuml/prims/cupy_helper.py` for now, and someone can shout if they think there's a better place for it.

Reply:
I changed my mind as I was working on this. `_imputation.py` already has similar helper functions in it, and I would want to make this function more general if we dropped it in something generic like `cuml/prims/cupy_helper.py`. I've pulled out all 3 of the masked calculations into their own functions and placed them with the other helpers in `_imputation.py` in my latest commit. @JohnZed, you happy with this organization?